File: pulse-android-otel/instrumentation/interaction/core/src/main/java/com/pulse/android/core/InteractionUtil.kt
package com.pulse.android.core

import com.pulse.android.remote.models.InteractionAttrsEntry
import com.pulse.android.remote.models.InteractionConfig
import com.pulse.android.remote.models.InteractionEvent
import com.pulse.utils.PulseOtelUtils
import java.util.Locale

internal data class InteractionBuildError(
    val type: InteractionErrorType,
    val timeoutExpectedEventName: String? = null,
    val sequenceViolationExpectedEventName: String? = null,
    val sequenceViolationReceivedEventName: String? = null,
)

internal object InteractionUtil {
    /**
     * Returns null when there is event of interest but because of ordering it didn't create any
     * change in the matching
     * For example global black listed event came but before or after the first or last event
     * respectively
     */
    @Suppress("LongMethod")
    fun matchSequence(
        ongoingMatchInteractionId: String,
        localEvents: List<InteractionLocalEvent>,
        localMarkers: List<InteractionLocalEvent>,
        interactionConfig: InteractionConfig,
    ): MatchResult? {
        val stepWiseTimeInNano: MutableList<InteractionLocalEvent> = mutableListOf()

        var configEventIndex = 0
        var isMatchOnGoing = false

        fun resetMatching() {
            stepWiseTimeInNano.clear()
            configEventIndex = 0
            isMatchOnGoing = false
        }

        var newInteractionStatus: MatchResult? = null
        logDebug { "localEvents = ${localEvents.joinToString { it.name }}" }
        var localEventIndex = 0
        while (localEventIndex < localEvents.size) {
            val localEvent = localEvents[localEventIndex]

            if (isMatchOnGoing && localEvent matchesAny interactionConfig.globalBlacklistedEvents) {
                logDebug { "blacklisted event(${localEvent.name}) found" }
                return MatchResult(
                    shouldTakeFirstEvent = false,
                    shouldResetList = true,
                    interactionStatus = InteractionRunningStatus.NoOngoingMatch(null),
                )
            }

            val configEvent = interactionConfig.events[configEventIndex]

            logDebug { "localEvent:${localEvent.name} from localEventIndex = $localEventIndex," }
            val isMatch = localEvent matches configEvent
            newInteractionStatus =
                if (isMatch) {
                    if (configEvent.isBlacklisted) {
                        logDebug { "localEvent:${localEvent.name} is blacklisted" }
                        MatchResult(
                            shouldTakeFirstEvent = false,
                            shouldResetList = true,
                            interactionStatus = InteractionRunningStatus.NoOngoingMatch(null),
                        )
                    } else {
                        stepWiseTimeInNano.add(localEvent)
                        configEventIndex++
                        logDebug {
                            "localEvent:${localEvent.name} is match and not a blacklisted match, " +
                                "matched at index = ${configEventIndex - 1}, " +
                                "config(w/o blacklisted) = ${interactionConfig.eventsSize}"
                        }

                        if (configEventIndex == interactionConfig.eventsSize) {
                            logDebug { "localEvent:${localEvent.name} is final match" }
                            isMatchOnGoing = false
                            MatchResult(
                                shouldTakeFirstEvent = false,
                                shouldResetList = true,
                                interactionStatus =
                                    InteractionRunningStatus.OngoingMatch(
                                        interactionId = ongoingMatchInteractionId,
                                        interactionConfig = interactionConfig,
                                        index = configEventIndex - 1,
                                        interaction =
                                            buildPulseInteraction(
                                                interactionId = ongoingMatchInteractionId,
                                                interactionConfig = interactionConfig,
                                                events = stepWiseTimeInNano,
                                                localMarkers = localMarkers,
                                            ),
                                    ),
                            )
                        } else {
                            isMatchOnGoing = true
                            // ongoing match
                            MatchResult(
                                shouldTakeFirstEvent = false,
                                shouldResetList = false,
                                interactionStatus =
                                    InteractionRunningStatus.OngoingMatch(
                                        index = configEventIndex - 1,
                                        interactionId = ongoingMatchInteractionId,
                                        interactionConfig = interactionConfig,
                                        interaction = null,
                                    ),
                            )
                        }
                    }
                } else if (configEvent.isBlacklisted) {
                    configEventIndex++
                    continue
                } else if (isMatchOnGoing) {
                    isMatchOnGoing = false
                    MatchResult(
                        shouldTakeFirstEvent = true,
                        shouldResetList = true,
                        interactionStatus =
                            InteractionRunningStatus.OngoingMatch(
                                index = configEventIndex - 1,
                                interactionId = ongoingMatchInteractionId,
                                interactionConfig = interactionConfig,
                                interaction =
                                    buildPulseInteraction(
                                        interactionId = ongoingMatchInteractionId,
                                        interactionConfig = interactionConfig,
                                        events = stepWiseTimeInNano,
                                        localMarkers = localMarkers,
                                        error =
                                            InteractionBuildError(
                                                type = InteractionErrorType.SEQUENCE_VIOLATION,
                                                sequenceViolationExpectedEventName = configEvent.name,
                                                sequenceViolationReceivedEventName = localEvent.name,
                                            ),
                                    ),
                            ),
                    )
                } else {
                    // no match is ongoing
                    null
                }
            localEventIndex++
        }

        if (newInteractionStatus?.shouldResetList == true) {
            resetMatching()
        }

        return newInteractionStatus
    }

    infix fun InteractionLocalEvent.matches(interactionEvent: InteractionEvent): Boolean {
        if (name != interactionEvent.name) return false
        val propsInteractionConfig = interactionEvent.props ?: return true
        val propsLocalEvent = this.props ?: return false
        return propsInteractionConfig.all { it in propsLocalEvent }
    }

    infix fun InteractionLocalEvent.matchesAny(interactionEvent: Iterable<InteractionEvent>) = interactionEvent.any { this matches it }

    private operator fun Map<String, String>.contains(propInteractionConfig: InteractionAttrsEntry): Boolean {
        val propName = propInteractionConfig.name
        val propValue = propInteractionConfig.value
        val operator = propInteractionConfig.operator

        if (!this.containsKey(propName)) return false

        val actualValue = this[propName]

        return matchPropValue(propValue, operator, actualValue)
    }

    private fun matchPropValue(
        expectedValue: String?,
        operator: String?,
        actualValue: String?,
    ): Boolean {
        if (expectedValue == null || operator == null || actualValue == null) return false
        val actualValueLower = actualValue.lowercase()
        val expectedValueLower = expectedValue.lowercase()

        return when (operator.uppercase(Locale.ROOT)) {
            InteractionConstant.Operators.EQUALS.operatorName -> actualValue == expectedValue
            InteractionConstant.Operators.NOT_EQUALS.operatorName -> actualValue != expectedValue
            InteractionConstant.Operators.CONTAINS.operatorName -> actualValueLower.contains(expectedValueLower)
            InteractionConstant.Operators.NOT_CONTAINS.operatorName -> !actualValueLower.contains(expectedValueLower)
            InteractionConstant.Operators.STARTS_WITH.operatorName -> actualValueLower.startsWith(expectedValueLower)
            InteractionConstant.Operators.ENDS_WITH.operatorName -> actualValueLower.endsWith(expectedValueLower)
            else -> false
        }
    }

    private fun interactionErrorMessage(error: InteractionBuildError): String =
        when (error.type) {
            InteractionErrorType.TIMEOUT -> {
                if (error.timeoutExpectedEventName != null) {
                    "Timed out while waiting for event \"${error.timeoutExpectedEventName}\"."
                } else {
                    "Timed out before the next expected event arrived."
                }
            }
            InteractionErrorType.SEQUENCE_VIOLATION -> {
                if (error.sequenceViolationExpectedEventName != null && error.sequenceViolationReceivedEventName != null) {
                    """Expected event \"${error.sequenceViolationExpectedEventName}\", received \"${error.sequenceViolationReceivedEventName}\"."""
                } else {
                    "An event did not match the next expected event in this interaction."
                }
            }
        }

    internal fun buildPulseInteraction(
        interactionId: String,
        interactionConfig: InteractionConfig,
        events: List<InteractionLocalEvent>,
        localMarkers: List<InteractionLocalEvent>,
        error: InteractionBuildError? = null,
    ): Interaction {
        require(events.isNotEmpty()) { "buildPulseInteraction requires at least one event" }
        val interactionName = interactionConfig.name
        val interactionConfigId = interactionConfig.id
        val lastEventTimeInNano = events.last().timeInNano
        val errorType = error?.type

        val errorMessage = error?.let { interactionErrorMessage(it) }

        val (timeDifferenceInNano, timeCategory, upTimeIndex) =
            if (errorType == null) {
                val timeDifferenceInNano = lastEventTimeInNano - events.first().timeInNano
                val timeDifferenceInMs = timeDifferenceInNano / 1000_000
                val lowerLimitInMs = interactionConfig.uptimeLowerLimitInMs
                val midLimitInMs = interactionConfig.uptimeMidLimitInMs
                val upperLimitInMs = interactionConfig.uptimeUpperLimitInMs

                val (upTimeIndex, timeCategory) =
                    when {
                        timeDifferenceInMs <= lowerLimitInMs -> {
                            1.0 to InteractionConstant.TimeCategory.EXCELLENT
                        }

                        timeDifferenceInMs <= midLimitInMs -> {
                            getUpTimeIndex(
                                timeDifferenceInMs,
                                lowerLimitInMs,
                                upperLimitInMs,
                            ) to InteractionConstant.TimeCategory.GOOD
                        }

                        timeDifferenceInMs <= upperLimitInMs -> {
                            getUpTimeIndex(
                                timeDifferenceInMs,
                                lowerLimitInMs,
                                upperLimitInMs,
                            ) to InteractionConstant.TimeCategory.AVERAGE
                        }

                        else -> {
                            0.0 to InteractionConstant.TimeCategory.POOR
                        }
                    }

                Triple(timeDifferenceInNano, timeCategory, upTimeIndex)
            } else {
                Triple(null, null, null)
            }
            
        val timeInMsDiffPair = events.getTimeSpanInNanos(interactionConfig.thresholdInMs)
        
        val maps =
            buildMap {
                put(InteractionConstant.NAME, interactionName)
                put(InteractionConstant.CONFIG_ID, interactionConfigId)
                put(InteractionConstant.LAST_EVENT_TIME_IN_NANO, lastEventTimeInNano)
                put(InteractionConstant.LOCAL_EVENTS, events.toList())
                put(InteractionConstant.MARKER_EVENTS, (
                    timeInMsDiffPair?.let { localMarkers.getEventsBetween(it.first, it.second) } ?: localMarkers.toList())
                )
                put(InteractionConstant.APDEX_SCORE, upTimeIndex)
                put(InteractionConstant.USER_CATEGORY, timeCategory?.categoryName)
                put(InteractionConstant.TIME_TO_COMPLETE_IN_NANO, timeDifferenceInNano)
                put(InteractionConstant.IS_ERROR, errorType != null)
                put(InteractionConstant.ERROR_TYPE, errorType?.code)
                put(InteractionConstant.ERROR_MESSAGE, errorMessage)
            }

        return Interaction(
            id = interactionId,
            name = interactionName,
            props = maps,
        )
    }

    private fun getUpTimeIndex(
        timeDifferenceInNano: Long,
        lowerLimit: Long,
        upperLimit: Long,
    ): Double = 1.0 - (1.0 * (timeDifferenceInNano - lowerLimit) / (upperLimit - lowerLimit))

    internal data class MatchResult(
        val shouldTakeFirstEvent: Boolean,
        val shouldResetList: Boolean,
        val interactionStatus: InteractionRunningStatus,
    )
}

internal inline fun logDebug(body: () -> String) {
    PulseOtelUtils.logDebug(InteractionConstant.LOG_TAG, body)
}

/**
 * Contains the info about generated interaction
 */
public class Interaction internal constructor(
    public val id: String,
    public val name: String,
    public val props: Map<String, Any?> = emptyMap(),
)

@Suppress("UNCHECKED_CAST")
internal fun Interaction.getTimeSpanInNanos(timeOutInMs: Long): Pair<Long, Long>? = events.getTimeSpanInNanos(timeOutInMs)

internal fun List<InteractionLocalEvent>.getTimeSpanInNanos(timeOutInMs: Long): Pair<Long, Long>? {
    val steps = this
    if (steps.isEmpty()) {
        PulseOtelUtils.logError(
            tag = InteractionConstant.LOG_TAG,
            throwable = IllegalStateException("getTimeSpanInNanos: Events size is 0)"),
        ) {
            "getTimeSpanInNanos: Events size is 0."
        }
        return null
    }
    val errorTypeParsed =
        (props[InteractionConstant.ERROR_TYPE] as? String)?.let { InteractionErrorType.fromCode(it) }
    if (errorTypeParsed != null) {
        val firstNs = steps.first().timeInNano
        val lastNs = steps.last().timeInNano
        val thresholdNs = timeOutInMs * 1_000_000L
        return firstNs to
            if (errorTypeParsed == InteractionErrorType.TIMEOUT) {
                firstNs + thresholdNs + (lastNs - firstNs)
            } else {
                lastNs
            }
    }
    if (steps.size == 1) {
        return steps[0].timeInNano to steps[0].timeInNano + timeOutInMs * 1000000
    }
    return steps.first().timeInNano to steps.last().timeInNano
}

internal fun List<InteractionLocalEvent>.getEventsBetween(
    startInNanoInclusive: Long,
    endInNanoInclusive: Long,
): List<InteractionLocalEvent> =
    this.filter {
        it.timeInNano in startInNanoInclusive..endInNanoInclusive
    }

public fun Interaction.getTimeSpanInNanos(interactionStatus: InteractionRunningStatus.OngoingMatch): Pair<Long, Long>? =
    this.getTimeSpanInNanos(timeOutInMs = interactionStatus.interactionConfig.thresholdInMs)

@Suppress("UNCHECKED_CAST")
public val Interaction.events: List<InteractionLocalEvent>
    get() {
        return props[InteractionConstant.LOCAL_EVENTS] as? List<InteractionLocalEvent>
            ?: error("InteractionConstant.LOCAL_EVENTS is missing or not of correct type")
    }

@Suppress("UNCHECKED_CAST")
public val Interaction.markerEvents: List<InteractionLocalEvent>
    get() {
        return props[InteractionConstant.MARKER_EVENTS] as? List<InteractionLocalEvent>
            ?: error("InteractionConstant.MARKER_EVENTS is missing or not of correct type")
    }

public val Interaction.isErrored: Boolean
    get() {
        val isError =
            props[InteractionConstant.IS_ERROR] as? Boolean
                ?: error("InteractionConstant.IS_ERROR is missing or not of correct type")
        return isError
    }

public val Interaction.errorTypeCode: String?
    get() = props[InteractionConstant.ERROR_TYPE] as? String

public val Interaction.errorMessage: String?
    get() = props[InteractionConstant.ERROR_MESSAGE] as? String

File: pulse-ios-otel/Sources/Instrumentation/Interaction/Core/InteractionUtil.swift
/*
 * Copyright The Pulse Authors
 * SPDX-License-Identifier: Apache-2.0
 */

import Foundation

/// Core utility for matching event sequences and building interactions
internal enum InteractionUtil {
    internal struct InteractionBuildError {
        let type: InteractionErrorType
        var timeoutExpectedEventName: String?
        var sequenceViolationExpectedEventName: String?
        var sequenceViolationReceivedEventName: String?
    }

    private static func interactionErrorMessage(_ error: InteractionBuildError) -> String {
        switch error.type {
        case .timeout:
            if let name = error.timeoutExpectedEventName {
                return "Timed out while waiting for event \"\(name)\"."
            }
            return "Timed out before the next expected event arrived."
        case .sequenceViolation:
            if let expected = error.sequenceViolationExpectedEventName,
               let received = error.sequenceViolationReceivedEventName {
                return "Expected event \"\(expected)\", received \"\(received)\"."
            }
            return "An event did not match the next expected event in this interaction."
        }
    }

    /// Result of matching a sequence
    struct MatchResult {
        let shouldTakeFirstEvent: Bool
        let shouldResetList: Bool
        let interactionStatus: InteractionRunningStatus
    }

    /**
     * Matches local events against interaction configuration sequence
     * Returns nil when there is event of interest but because of ordering it didn't create any
     * change in the matching
     * For example global black listed event came but before or after the first or last event
     * respectively
     */
    static func matchSequence(
        ongoingMatchInteractionId: String,
        localEvents: [InteractionLocalEvent],
        localMarkers: [InteractionLocalEvent],
        interactionConfig: InteractionConfig
    ) -> MatchResult? {
        var stepWiseTimeInNano: [InteractionLocalEvent] = []
        var configEventIndex = 0
        var isMatchOnGoing = false

        func resetMatching() {
            stepWiseTimeInNano.removeAll()
            configEventIndex = 0
            isMatchOnGoing = false
        }

        var newInteractionStatus: MatchResult?

        var localEventIndex = 0
        while localEventIndex < localEvents.count {
            let localEvent = localEvents[localEventIndex]

            // Check for global blacklisted events during ongoing match
            if isMatchOnGoing && matchesAny(localEvent, interactionConfig.globalBlacklistedEvents) {
                return MatchResult(
                    shouldTakeFirstEvent: false,
                    shouldResetList: true,
                    interactionStatus: .noOngoingMatch(oldOngoingInteractionRunningStatus: nil)
                )
            }

            let configEvent = interactionConfig.events[configEventIndex]
            let isMatch = matches(localEvent, configEvent)

            if isMatch {
                if configEvent.isBlacklisted {
                    newInteractionStatus = MatchResult(
                        shouldTakeFirstEvent: false,
                        shouldResetList: true,
                        interactionStatus: .noOngoingMatch(oldOngoingInteractionRunningStatus: nil)
                    )
                } else {
                    stepWiseTimeInNano.append(localEvent)
                    configEventIndex += 1

                    if configEventIndex == interactionConfig.eventsSize {
                        isMatchOnGoing = false
                        newInteractionStatus = MatchResult(
                            shouldTakeFirstEvent: false,
                            shouldResetList: true,
                            interactionStatus: .ongoingMatch(
                                index: configEventIndex - 1,
                                interactionId: ongoingMatchInteractionId,
                                interactionConfig: interactionConfig,
                                interaction: buildPulseInteraction(
                                    interactionId: ongoingMatchInteractionId,
                                    interactionConfig: interactionConfig,
                                    events: stepWiseTimeInNano,
                                    localMarkers: localMarkers,
                                    error: nil
                                )
                            )
                        )
                    } else {
                        isMatchOnGoing = true
                        // ongoing match
                        newInteractionStatus = MatchResult(
                            shouldTakeFirstEvent: false,
                            shouldResetList: false,
                            interactionStatus: .ongoingMatch(
                                index: configEventIndex - 1,
                                interactionId: ongoingMatchInteractionId,
                                interactionConfig: interactionConfig,
                                interaction: nil
                            )
                        )
                    }
                }
            } else if configEvent.isBlacklisted {
                // Skip the blacklisted config event, but keep checking the current local event
                // against the next config event
                configEventIndex += 1
                continue
            } else if isMatchOnGoing {
                isMatchOnGoing = false
                newInteractionStatus = MatchResult(
                    shouldTakeFirstEvent: true,
                    shouldResetList: true,
                    interactionStatus: .ongoingMatch(
                        index: configEventIndex - 1,
                        interactionId: ongoingMatchInteractionId,
                        interactionConfig: interactionConfig,
                        interaction: buildPulseInteraction(
                            interactionId: ongoingMatchInteractionId,
                            interactionConfig: interactionConfig,
                            events: stepWiseTimeInNano,
                            localMarkers: localMarkers,
                            error: InteractionBuildError(
                                type: .sequenceViolation,
                                sequenceViolationExpectedEventName: configEvent.name,
                                sequenceViolationReceivedEventName: localEvent.name
                            )
                        )
                    )
                )
            } else {
                // no match is ongoing
                // newInteractionStatus remains nil
            }

            localEventIndex += 1
        }

        if newInteractionStatus?.shouldResetList == true {
            resetMatching()
        }

        return newInteractionStatus
    }

    /// Check if local event matches interaction event
    static func matches(_ localEvent: InteractionLocalEvent, _ interactionEvent: InteractionEvent) -> Bool {
        if localEvent.name != interactionEvent.name {
            return false
        }

        guard let propsInteractionConfig = interactionEvent.props else {
            return true
        }

        guard let propsLocalEvent = localEvent.props else {
            return false
        }

        return propsInteractionConfig.allSatisfy { prop in
            matchesProperty(prop, in: propsLocalEvent)
        }
    }

    /// Check if local event matches any of the interaction events
    static func matchesAny(_ localEvent: InteractionLocalEvent, _ interactionEvents: [InteractionEvent]) -> Bool {
        return interactionEvents.contains { matches(localEvent, $0) }
    }

    /// Check if property matches in the local event props
    private static func matchesProperty(_ propInteractionConfig: InteractionAttrsEntry, in propsLocalEvent: [String: String]) -> Bool {
        let propName = propInteractionConfig.name
        let propValue = propInteractionConfig.value
        let operatorValue = propInteractionConfig.operator

        guard let actualValue = propsLocalEvent[propName] else {
            return false
        }

        return matchPropValue(expectedValue: propValue, operator: operatorValue, actualValue: actualValue)
    }

    /// Match property value using operator
    private static func matchPropValue(
        expectedValue: String,
        operator: String,
        actualValue: String
    ) -> Bool {
        let actualValueLower = actualValue.lowercased()
        let expectedValueLower = expectedValue.lowercased()
        let operatorUpper = `operator`.uppercased()

        guard let op = InteractionAttributes.Operators(rawValue: operatorUpper) else {
            return false
        }

        switch op {
        case .equals:
            return actualValue == expectedValue
        case .notEquals:
            return actualValue != expectedValue
        case .contains:
            return actualValueLower.contains(expectedValueLower)
        case .notContains:
            return !actualValueLower.contains(expectedValueLower)
        case .startsWith:
            return actualValueLower.hasPrefix(expectedValueLower)
        case .endsWith:
            return actualValueLower.hasSuffix(expectedValueLower)
        }
    }

    /// Build the final Interaction object
    static func buildPulseInteraction(
        interactionId: String,
        interactionConfig: InteractionConfig,
        events: [InteractionLocalEvent],
        localMarkers: [InteractionLocalEvent],
        error: InteractionBuildError? = nil
    ) -> Interaction {
        guard let firstEvent = events.first, let lastEvent = events.last else {
            preconditionFailure("buildPulseInteraction requires at least one event")
        }

        let interactionName = interactionConfig.name
        let interactionConfigId = interactionConfig.id
        let lastEventTimeInNano = lastEvent.timeInNano

        let (timeDifferenceInNano, timeCategory, upTimeIndex): (Int64?, InteractionAttributes.TimeCategory?, Double?)

        if error == nil {
            let timeDiffInNano = lastEvent.timeInNano - firstEvent.timeInNano
            let timeDifferenceInMs = timeDiffInNano / 1_000_000
            let lowerLimitInMs = interactionConfig.uptimeLowerLimitInMs
            let midLimitInMs = interactionConfig.uptimeMidLimitInMs
            let upperLimitInMs = interactionConfig.uptimeUpperLimitInMs

            let (upTimeIdx, timeCat): (Double, InteractionAttributes.TimeCategory)
            if timeDifferenceInMs <= lowerLimitInMs {
                upTimeIdx = 1.0
                timeCat = .excellent
            } else if timeDifferenceInMs <= midLimitInMs {
                upTimeIdx = getUpTimeIndex(
                    timeDifferenceInMs: timeDifferenceInMs,
                    lowerLimit: lowerLimitInMs,
                    upperLimit: upperLimitInMs
                )
                timeCat = .good
            } else if timeDifferenceInMs <= upperLimitInMs {
                upTimeIdx = getUpTimeIndex(
                    timeDifferenceInMs: timeDifferenceInMs,
                    lowerLimit: lowerLimitInMs,
                    upperLimit: upperLimitInMs
                )
                timeCat = .average
            } else {
                upTimeIdx = 0.0
                timeCat = .poor
            }

            (timeDifferenceInNano, timeCategory, upTimeIndex) = (timeDiffInNano, timeCat, upTimeIdx)
        } else {
            (timeDifferenceInNano, timeCategory, upTimeIndex) = (nil, nil, nil)
        }

        let maps: [String: Any?] = {
            var m: [String: Any?] = [
                InteractionAttributes.name: interactionName,
                InteractionAttributes.configId: interactionConfigId,
                InteractionAttributes.lastEventTimeInNano: lastEventTimeInNano,
                InteractionAttributes.localEvents: events,
                InteractionAttributes.markerEvents: localMarkers,
                InteractionAttributes.apdexScore: upTimeIndex,
                InteractionAttributes.userCategory: timeCategory?.rawValue,
                InteractionAttributes.timeToCompleteInNano: timeDifferenceInNano,
                InteractionAttributes.isError: error != nil
            ]
            if let error = error {
                m[InteractionAttributes.errorType] = error.type.code
                m[InteractionAttributes.errorMessage] = interactionErrorMessage(error)
            }
            return m
        }()

        return Interaction(
            id: interactionId,
            name: interactionName,
            props: maps
        )
    }

    /// Calculate uptime index for APDEX score
    private static func getUpTimeIndex(
        timeDifferenceInMs: Int64,
        lowerLimit: Int64,
        upperLimit: Int64
    ) -> Double {
        guard upperLimit > lowerLimit else {
            return 0.0
        }
        return 1.0 - (1.0 * Double(timeDifferenceInMs - lowerLimit) / Double(upperLimit - lowerLimit))
    }
}
